File: r-poly-regs-preds.rb
#!/usr/bin/ruby

# Examples: /usr/lib/ruby/gems/1.8/gems/rsruby-0.5.1.1/examples
# Examples: :~/R/i686-pc-linux-gnu-library/2.13/quantreg/demo
# http://rubyforscientificresearch.blogspot.com/search?q=RSruby

# this example shows how to pass data between Ruby and R
# Regression followed by a prediction based on the regression
# Both mse and quantile regressions are in the example
# quantile = nil does mse and quantile between 0 and 1 does quantile regression

######################### input data ###############################
imax = 20         # number of data points
iahead = 2        # points to predict past imax
degree = 2        # polynomial degree
level = 0.95      # regression confidence level                                
quantile = nil   # regression type:  nil for mse, between 0 and 1 for quantile
print = true      # true for verbose

puts "\nParameters::\n\tdata points: #{imax}\n\tprediction points: #{iahead}\n\tdegree: #{degree}\n\tconfidence: #{level}\n\tquantile: #{quantile.inspect}"

################################ Load Gems ###########################

ENV['R_HOME'] = "/usr/lib/R"  # need to define the R_HOME

# some basic error checking when loading gems
def load_gem(a_gem)
  begin
    require a_gem
  rescue LoadError
    $stderr.print "#{a_gem} missing. Exiting"
    exit
  end
end

load_gem('rubygems')
load_gem('rsruby')
r = RSRuby.instance 

#######################################################################

######################### FUNCTIONS ###################################

# this is needed to put the fit coefficients as the order is not guaranteed
def get_coefs(mdl, degree, coefs)
  kcoefs = [coefs["(Intercept)"]]                 # polynomial coefficient keys
  if degree == 1                                  # set polynomial coefficient keys
    kcoefs[1] = coefs[mdl]
  else
    (1..degree).each {|i| kcoefs[i] = coefs[mdl + i.to_s]}
  end
  return kcoefs
end

def poly_val(coefs, x)
  y, xp = 0, 1.0
  coefs.each do |c|
    y += c * xp
    xp *= x
  end
  return y
end

def print_coefs(smry, size, print)  # prints coef stats and returns their values
  vcoefs = smry["coefficients"]
  coefs, upr_coefs, lwr_coefs = [], [], []
  puts "\tCoefficients\n\tDegree\t\tEstimate\t\t\tError\t\t\t\ttValue\t\t\t\tProb >|t|" if print == true
  (0...size).each do |d|
    puts "\t#{d}\t\t#{vcoefs[d][0]}\t\t#{vcoefs[d][1]}\t\t#{vcoefs[d][2]}\t\t#{vcoefs[d][3]}" if print == true
    coefs << vcoefs[d][0]
    upr_coefs << vcoefs[d][0] + vcoefs[d][1]
    lwr_coefs << vcoefs[d][0] - vcoefs[d][1]
  end
  return coefs, upr_coefs, lwr_coefs
end

def set_data(imax, iahead, d)
  x, y = [], [] 
  (0...imax).each do |i|
    x << i
    y << 2 * rand() * (x[-1] ** d) + rand()
  end
  return x, y
end

def do_pred(degree, x, y, level, r, imax, iahead, quantile, print)
  # assign to R objects to input data
  r.assign('xr', x)
  r.assign('yr', y)
  dt = r.eval_R("dt <- data.frame(xr,yr)")
  
  ################# model ###################
  s = quantile == nil ? "\nMSE Regression" : "\nQuantile Regression"
  puts s          # type of regression 
  r.assign('rsp', "response")
  r.assign('prd', "prediction")
  r.assign('se',"boot")  # boot, ker, nid, iid, rank. Only boot or iid work. needed for quantile
  r.assign('t', "none")       # needed for quantile
  r.assign('c',"confidence")  # needed for quantil
  model = "poly(xr, #{degree}, raw = T)"    # regression model
  
  if quantile != nil      # quantile regression
    lib = "library(\'quantreg\')"
    r.eval_R("suppressPackageStartupMessages(#{lib})")          
    r.assign('quant', quantile)
    fit = r.eval_R("mdl <- rq(formula = yr ~ #{model}, tau = quant, data = dt, model = TRUE)")
    smry = r.eval_R("smry <- summary(mdl,se = se, level = #{level})")
    pred1 = r.eval_R("p1 <- predict(mdl, newdata = dt, type = t, level = #{level}, interval = c, se = se)")  # works only for se = boot, iid, 
  else                      # mse regression
    fit = r.eval_R("mdl <- lm(formula = yr ~ #{model},data = dt)") 
    smry = r.eval_R("smry <- summary(mdl,se = se, level = #{level})")
    pred1 = r.eval_R("p1 <- predict(mdl, se = TRUE, newdata = dt, type = rsp, level = #{level}, interval = prd)")
  end
  
  coefs, upr_coefs, lwr_coefs = print_coefs(smry, 1 + degree, print)  # print and extract coefs
  puts "\nF-statistic: #{smry["fstatistic"]["value"].inspect}" if quantile == nil and print == true
  
  ################### Model validation #########################
  puts "\n\nData and Model Fit" if print == true
  puts "\ty: actual data" if print == true
  puts "\tyh: model data directly computed." if print == true
  puts "\tyR, yRupr, yRlwr: model data computed by R" if print == true
  (0...imax).each do |i|
    yh = poly_val(coefs, x[i])
    if quantile == nil          # mse
      yR, yRlwr, yRupr = pred1["fit"][i]
      se = pred1["se.fit"][(1+i).to_s]
    else                        # quantile
      yR, yRlwr, yRupr = pred1[i]
      se = (yRupr - yRlwr) / 2.0
    end
    puts "\t#{i}  x: #{x[i]}  y: #{y[i]}  yh: #{yh} yR: #{yR} yRlwr: #{yRlwr} yRupr: #{yRupr} se: #{se}" if print == true
  end
  
  ################# Prediction ##################
  x = []
  iahead = degree + 1 if degree <= iahead        # predict() must have at more points to predict than the poly degree (R issues)
  (imax...imax + iahead).each {|idx| x << idx}
  r.assign('xr', x)
  ndt = r.eval_R("ndt <- data.frame(xr)") # define the prediction data range
  if quantile == nil    # mse
    pred2 = r.eval_R("p2 <- predict(mdl, newdata = ndt, se = TRUE, type = rsp, level = #{level}, interval = prd)")
  else                  # quantile
    pred2 = r.eval_R("p2 <- predict(mdl, newdata = ndt, type = t, level = #{level}, interval = c, se = se)")
  end
 
  puts "\nPredictions"
  (imax...imax + iahead).each do |idx|
    xval = idx
    if quantile == nil    # mse
      yR, yRlwr, yRupr = pred2['fit'][idx - imax]
      se = pred2["se.fit"][(1 + idx - imax).to_s]
    else    # quantile
      yR, yRlwr, yRupr = pred2[idx  - imax]
      se = (yRupr - yRlwr) / 2.0
    end
    #yh = poly_val(coefs, xval)
    puts "\t#{idx}  x: #{xval}  y: N/A  yR: #{yR} yRlwr: #{yRlwr} yRupr: #{yRupr} se: #{se}"
  end
end

########################################################################################################

############################################## MAIN ####################################################
x, y = set_data(imax, iahead, degree)
do_pred(degree, x, y, level, r, imax, iahead, quantile, print)
puts "Done"

########################################################################################################

File: r-poly-regs-preds.py
#! /usr/bin/env python

############################### Parameters for regressions #################################

points = 100        # number of data points
prediction = 2      # prediction points ahead
degree = 2          # polynomial degree
quantile = 0.75     # between 0 and 1
conf = 0.95         # conf level for regression
print "Input:: points: %d degree: %d quantile: %f conf level: %f" %(points, degree, quantile, conf)

##################################### Python imports ###################################################

import sys
import rpy2.robjects as robjects
import math as math

rObj = robjects.r
stats = robjects.packages.importr(name = 'stats', suppress_messages = True)     # load stats package
base = robjects.packages.importr(name = 'base', suppress_messages = True)       # load stats package
qreg = robjects.packages.importr(name = 'quantreg', suppress_messages = True)   # load quantreg package

########################################################################################################

############################################ Generate data ##############################################
x, y = [], []
for idx in range(0, points):
    x.append(idx)
    z = rObj.rnorm(1, 0, 10)
    y.append(idx ** (2 + degree) + z[0])  # anything here really

xr = robjects.FloatVector(x)
yr = robjects.FloatVector(y)

#          
# prediction data
#
xpred = []
if degree <= prediction:        # predict() must have at more points to predict than the poly degree (R issues)
    prediction = degree + 1
    print "\nPREDICTION: prediction increased to %d to avoid R issue" %prediction
for idx in range(points, points + prediction):
    xpred.append(idx)
xpredr = robjects.FloatVector(xpred)
newdata = base.data_frame(xr = xpredr)

#########################################################################################################

############################################## Formula ##################################################
model = 'poly(xr, %s, raw = T)'   %(str(degree)) 
fmla = robjects.Formula('yr ~ %s' % model)
env = fmla.environment
env['xr'] = xr
env['yr'] = yr

#
# MSE Regression
#
print "\n############## MSE REGRESSION ####################"
lmfit = stats.lm(fmla)
smry = base.summary(lmfit)
print "F statistic: %f" %smry.rx('fstatistic')[0][0]
print "R squared: %f" %smry.rx('r.squared')[0][0]
print "Adj R squared: %f" %smry.rx('adj.r.squared')[0][0]
print "Stdev: %f" %smry.rx('sigma')[0][0]
print '\n'
print "Coefficients"
coefs = smry.rx('coefficients')
print coefs
print 'Covariances'
covs = smry.rx('cov.unscaled')
print covs
print "\nPrediction"
pred = stats.predict(lmfit, newdata = newdata, se = True, interval = 'prediction', type = 'response', level = conf)
print pred.rx('fit')
print pred.rx('se.fit')
        
#
# quantile regression
#
print '\n######################## %s QUANTILE REGRESSION #####################' %quantile
qreg = robjects.packages.importr(name = 'quantreg', suppress_messages = True)   # load quantreg package
qfit = qreg.rq(fmla, tau = quantile, model = True)
smry = base.summary(qfit, se = 'boot', level = conf)     # se = boot or iid

print "Coefficients"
coefs = smry.rx('coefficients')
print coefs

print "\nPrediction"
# for quantile prediction must use:
#     type = none
#     se = boot or iid
#     interval = confidence
# Must use direct assignments to make it work

rObj.assign('xr', xpredr)
rObj('ndt = data.frame(x = xr)')
rObj.assign('s', 'boot')
rObj.assign('t', 'none')
rObj.assign('c', 'confidence')
rObj.assign('conf', conf)
rObj.assign('qfit', qfit)
pred = rObj('pred = predict(qfit, newdata = ndt, se = s, interval = c, type = t, level = conf)')
print pred
print "Done"
